rtos/Mutex.cpp

```diff
@@ -27,7 +27,7 @@
 
 namespace rtos {
 
-Mutex::Mutex()
+Mutex::Mutex(): _count(0)
```

Comment:
As I expressed in a previous comment, I think having public "get_count" is dangerous, as it's a programming trap. No correctly-written code needs it, and recursive mutex implementations don't conventionally provide it, which is why it's being manually tracked here.

The current `ConditionVariable` implementation doesn't really need it - it's there only to support a helpful `MBED_ASSERT` - detecting the programming error of calling `ConditionVariable:wait` with a count other than 1.

@c1728p9 countered my reservation about having `get_count` by noting that it was exposed only via `friend`. (I think a previous version had a private getter function, and still with `friend`?).

My personal preference would be to avoid the friend thing, and make it public, but conditional on `!NDEBUG` to make clear that it is a for-diagnostic-only API. I can see being able to access that information is useful for diagnostics and debugging, but really don't want any danger of any real code trying to do anything based on the count. Or maybe we could provide it but "deprecated" thing if `NDEBUG`?

The first draft of `ConditionVariable` used the count by "helpfully" unlocking and relocking the mutex N times where N was the current count, but that is not a good idea, as explained here: https://groups.google.com/forum/#!msg/comp.programming.threads/B0JE0_b50Ww/8ekMG09xewsJ

That feature was the original reason for adding the mutex count, but now it's just for that assert.

Reply:
My concern was regarding having related but possibly logically separated changes in one commit without any background (your comment helps now to get better picture, thanks).

It is fine as it is as stated above, should not hold this any longer

Reply:
Just wanted to push back on the getter method. Would agree on separating the commit.

Reply:
I can move this to a separate commit if you would like but I don't really see the purpose of this. Without ConditionVariable `_count` is incremented and decremented, but never used and cannot be used as it is private. Furthermore, it isn't even helpful for git operations, as reverting it breaks condition variable. Alternatively reverting ConditionVariable leaves unused code. @0xc0170  let me know if you still want me to split this into two commits, as testing will need to be rerun.